test/spoof/spoof.go

```diff
@@ -82,72 +85,81 @@ type SpoofingClient struct {
 	RequestInterval time.Duration
 	RequestTimeout  time.Duration
 
-	endpoint string
-	domain   string
-
 	logf logging.FormatLogger
 }
 
-// New returns a SpoofingClient that rewrites requests if the target domain is not `resolveable`.
+// TransportOption allows callers to customize the http.Transport used by a SpoofingClient
+type TransportOption func(transport *http.Transport) *http.Transport
+
+// New returns a SpoofingClient that rewrites requests if the target domain is not `resolvable`.
 // It does this by looking up the ingress at construction time, so reusing a client will not
 // follow the ingress if it moves (or if there are multiple ingresses).
 //
 // If that's a problem, see test/request.go#WaitForEndpointState for oneshot spoofing.
-func New(kubeClientset *kubernetes.Clientset, logf logging.FormatLogger, domain string, resolvable bool, endpointOverride string) (*SpoofingClient, error) {
-	sc := SpoofingClient{
-		Client:          &http.Client{Transport: &ochttp.Transport{Propagation: &b3.HTTPFormat{}}}, // Using ochttp Transport required for zipkin-tracing
-		RequestInterval: requestInterval,
-		RequestTimeout:  RequestTimeout,
-		logf:            logf,
+func New(
+	kubeClientset *kubernetes.Clientset,
+	logf logging.FormatLogger,
+	domain string,
+	resolvable bool,
+	endpointOverride string,
+	opts ...TransportOption) (*SpoofingClient, error) {
+	endpoint, err := ResolveEndpoint(kubeClientset, domain, resolvable, endpointOverride)
+	if err != nil {
+		fmt.Errorf("failed get the cluster endpoint: %v", err)
 	}
 
-	var err error
-	if sc.endpoint, err = ResolveEndpoint(kubeClientset, domain, resolvable, endpointOverride); err != nil {
-		return nil, err
+	// Spoof the hostname at the resolver level
+	dialContext := (&net.Dialer{}).DialContext
```

Comment:
probably can be a package level var, since `DialContext` is safe for concurrent usage